[VDG] Prevent window closing when dialog is open

wieslawsoltes opened the pull request:
Fixes #9983

https://user-images.githubusercontent.com/2297442/222978483-0f1dbcc3-1ca4-40ea-8edf-f371b1065056.mp4

yahiheb reviewed (approved):
tACK

yahiheb commented:
FTR this fixes the first issue in #9402

soosr commented:
Honestly, this feels super bad... There should be at least a confirmation dialog about the closing instead of completely ignoring the action. Like, do you want to close the application? Cancel/Yes.

An alternative solution could be to update the layout of the dialogs. Move the cancel to the top right. As it seems to me from the issue https://github.com/zkSNACKs/WalletWasabi/issues/9402 that the button was mistakenly used. This layout would make it clear what closes what.
![image](https://user-images.githubusercontent.com/16364053/206126358-11c52fb0-fd50-47d5-808b-e04f4279163f.png)

yahiheb commented:
Or simply block the main wallet view if a dialog is open so the user cannot click the X button.

ichthus1604 commented:
@soosr both Avalonia and WPF standard `Window.ShowDialog()` method actually prevents any interaction with the underlying (owner) window, so you can't even click minimize/close button, you need to close dialog first. I didn't test on Winforms but I think it's the same. 

IMHO since we're not using standard dialog mechanism, we should try to maintain standard UI behavior as much as possible.

wieslawsoltes commented:
> Honestly, this feels super bad... There should be at least a confirmation dialog about the closing instead of completely ignoring the action. Like, do you want to close the application? Cancel/Yes.
> 
> An alternative solution could be to update the layout of the dialogs. Move the cancel to the top right. As it seems to me from the issue #9402 that the button was mistakenly used. This layout would make it clear what closes what. ![image](https://user-images.githubusercontent.com/16364053/206126358-11c52fb0-fd50-47d5-808b-e04f4279163f.png)



> @soosr both Avalonia and WPF standard `Window.ShowDialog()` method actually prevents any interaction with the underlying (owner) window, so you can't even click minimize/close button, you need to close dialog first. I didn't test on Winforms but I think it's the same.
> 
> IMHO since we're not using standard dialog mechanism, we should try to maintain standard UI behavior as much as possible.
> 
> Ideally I would completely disable the title bar icons, but I couldn't find an Avalonia API that does this.

So I like the ideas:
1. Disable title bar buttons while dialog is visible
2. Add close button to visible top-most dialog

Will try to do that and see how it works.

yahiheb commented:
Why not simply block the main wallet view if a dialog is open so the user cannot click the X button.

This is imho too much.

![Capture](https://user-images.githubusercontent.com/52379387/218636262-f93c84fd-6b44-4d88-b6e4-b8f6d1e26e23.PNG)

soosr commented:
@wieslawsoltes Fix conflict please

yahiheb commented:
> > Why not simply block the main wallet view if a dialog is open so the user cannot click the X button.
> 
> Somehow the user has to know what is the problem. Marking the dialog with something is a good idea.

My point was to disable the buttons to minimize or close Wasabi (X button), so the user won't even deal with any of this or be notified, they would simply understand that they should first close the other dialog then close Wasabi.

soosr commented:
> My point was to disable the buttons to minimize or close Wasabi (X button)

Hard to do it cross platform.

wieslawsoltes commented:
> > My point was to disable the buttons to minimize or close Wasabi (X button)
> 
> Hard to do it cross platform.

Especially Linux and probably macOS

wieslawsoltes commented:
> I agree with this. I would use the secondary accent color (blue) for the animation. And make the animation more like a blinking one. Faster, and last only for a second.

Refactored to show only once and changed color.

soosr commented:
Could you fix the conflict?

yahiheb commented:
> Could you fix the conflict?

Fixed.

yahiheb reviewed (dismissed):
tACK

yahiheb commented:
This fixes #9983, so I added that to the PR description.

soosr commented on WalletWasabi.Fluent/ApplicationStateManager.cs:
```diff
@@ -126,7 +126,7 @@ private void CreateAndShowMainWindow()
 
 		Observable.FromEventPattern<CancelEventArgs>(result, nameof(result.Closing))
-			.Select(args => (args.EventArgs, !ApplicationViewModel.CanShutdown()))
+			.Select(args => (args.EventArgs, !ApplicationViewModel.CanShutdown(), ApplicationViewModel.MainViewCanShutdown()))
```
passing `ApplicationViewModel.MainViewCanShutdown()` doesn't seem necessary.

wieslawsoltes replied:
It's leftover, removed

soosr commented on WalletWasabi.Fluent/ViewModels/ApplicationViewModel.cs:
```diff
@@ -67,11 +67,37 @@ private IObservable<bool> AboutCanExecute()
+
+	public bool MainViewCanShutdown()
+	{
+		if (MainViewModel.Instance.IsDialogOpen())
+		{
+			return false;
+		}
+
+		return true;
```
```suggestion
		return !MainViewModel.Instance.IsDialogOpen();
```

wieslawsoltes replied:
Done

soosr commented on WalletWasabi.Fluent/ViewModels/ApplicationViewModel.cs:
```diff
@@ -67,11 +67,37 @@ private IObservable<bool> AboutCanExecute()
 
 	public void OnShutdownPrevented(bool restartRequest)
 	{
+		if (!MainViewCanShutdown())
+		{
+			MainViewModel.Instance.ShowDialogAlert();
+			return;
+		}
```
This should be placed after 
```c#
MainViewModel.Instance.ApplyUiConfigWindowState(); // Will pop the window if it was minimized.
```
Because the window should be brought to the front when it is minimized.

BTW question... Should we block the user when they try to close the software from the tray?

wieslawsoltes replied:
Fixed.

Not sure about blocking tray exit, guess we should be consistent and block all exits ?

soosr commented on WalletWasabi.Fluent/ViewModels/MainViewModel.cs:
```diff
@@ -135,6 +135,36 @@ public MainViewModel()
+
+		if (FullScreen.IsDialogOpen)
+		{
+			DialogScreen.ShowAlert = false;
```
Copy paste mistake, It should be `FullScreen`-

soosr replied:
BTW does the full screen even make any sense? Does the warning even animation visible when it is full screen?

wieslawsoltes replied:
Fixed. 

Animation is visible in Add Wallet screen.

soosr commented on WalletWasabi.Fluent/Controls/Dialog.axaml.cs:
```diff
@@ -216,13 +225,25 @@ protected override void OnPropertyChanged<T>(AvaloniaPropertyChangedEventArgs<T>
+
+			PseudoClasses.Set(":open", isOpen);
+
+			if (!isOpen)
+			{
+				PseudoClasses.Set(":alert", false);
+			}
```
Is this necessary? 

In the `Close()` method, it is set to false so `:alert` would be set to false anyway.

wieslawsoltes replied:
IsDialogOpen property is bound to view model so we need to handle can when property changed.

wieslawsoltes replied:
Yes

https://github.com/zkSNACKs/WalletWasabi/blob/a62e10e552d9a392ece5c3aaff275ce50fa08579/WalletWasabi.Fluent/ViewModels/Dialogs/Base/DialogScreenViewModel.cs#L32

https://github.com/zkSNACKs/WalletWasabi/blob/a62e10e552d9a392ece5c3aaff275ce50fa08579/WalletWasabi.Fluent/ViewModels/Dialogs/Base/DialogScreenViewModel.cs#L45

soosr reviewed (commented):
First round, will continue

soosr commented on WalletWasabi.Fluent/Controls/Dialog.axaml:
```diff
@@ -56,12 +56,34 @@
     <Setter Property="BoxShadow" Value="{DynamicResource DialogBoxShadow}" />
   </Style>
+
+  <Style Selector="c|Dialog:alert /template/ Border#PART_Border">
+    <Setter Property="BoxShadow" Value="{DynamicResource DialogBoxShadow}"/>
+    <Style.Animations>
+      <Animation Duration="0:0:0.8" PlaybackDirection="AlternateReverse" IterationCount="1">
+        <KeyFrame Cue="0%">
+          <Setter Property="BoxShadow" Value="{DynamicResource DialogAlertStartBoxShadow}"/>
+        </KeyFrame>
+        <KeyFrame Cue="25%">
+          <Setter Property="BoxShadow" Value="{DynamicResource DialogAlertEndBoxShadow}"/>
+        </KeyFrame>
+        <KeyFrame Cue="75%">
+          <Setter Property="BoxShadow" Value="{DynamicResource DialogAlertStartBoxShadow}"/>
+        </KeyFrame>
+        <KeyFrame Cue="100%">
+          <Setter Property="BoxShadow" Value="{DynamicResource DialogBoxShadow}"/>
+        </KeyFrame>
+      </Animation>
+    </Style.Animations>
+  </Style>
```
I would adjust the animation a tiny bit, it feels too slow. It might be only a personal preference so only apply this if someone else agrees too.

Current one:
![dialog_default](https://user-images.githubusercontent.com/16364053/222668094-d45ff11d-f962-48e2-b385-19533ce5d1fa.gif)

Adjusted:
![dialog_adjusted](https://user-images.githubusercontent.com/16364053/222668128-1ae5e5b6-07d7-4019-9427-201f96113b12.gif)

```xml
<Animation Duration="0:0:0.6" IterationCount="1">
  <KeyFrame Cue="25%">
    <Setter Property="BoxShadow" Value="{DynamicResource DialogAlertStartBoxShadow}"/>
  </KeyFrame>
  <KeyFrame Cue="50%">
    <Setter Property="BoxShadow" Value="{DynamicResource DialogAlertEndBoxShadow}" />
  </KeyFrame>
  <KeyFrame Cue="75%">
    <Setter Property="BoxShadow" Value="{DynamicResource DialogAlertStartBoxShadow}"/>
  </KeyFrame>
  <KeyFrame Cue="100%">
    <Setter Property="BoxShadow" Value="{DynamicResource DialogBoxShadow}"/>
  </KeyFrame>
</Animation>
```

Start BoxShadow: 0 0 10 5
End BoxShadow: 0 0 10 1

wieslawsoltes replied:
Adjusted animation and box shadow

soosr commented on WalletWasabi.Fluent/Controls/Dialog.axaml.cs:
```diff
@@ -237,15 +258,40 @@ protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
 			topLevel.AddHandler(KeyDownEvent, CancelKeyDown, RoutingStrategies.Tunnel);
 		}
+
+		AddHandler(PointerPressedEvent, HandlePointerPressed, RoutingStrategies.Tunnel);
+		AddHandler(KeyDownEvent, HandleKeyDown, RoutingStrategies.Tunnel);
```
My point is there are many `ShowAlert = false;` occurrences which makes me think that something should be simplified.

yahiheb reviewed (approved):
Tested, LGTM.

soosr reviewed (commented):
nit

wieslawsoltes commented:
> nit

Committed proposed change

wieslawsoltes commented:
> tACK
> 
> (For some reason the first few animations are different, it looks like they don't complete. I think it is not related to the PR)

Yeah might be animation system, hopefully with v11 we can rewrite using compositor animations and run on render thread so UI is not affected that much.

yahiheb reviewed (approved):
tACK
